runtime/lua/vim/keymap.lua

```diff
@@ -70,13 +71,16 @@ function keymap.set(mode, lhs, rhs, opts)
     rhs = ''
   end
 
-  if opts.buffer then
-    local bufnr = opts.buffer == true and 0 or opts.buffer --[[@as integer]]
+  opts.buf = opts.buffer or opts.buf
+  if opts.buf then
+    local bufnr = opts.buf == true and 0 or opts.buf --[[@as integer]]
+    opts.buf = nil ---@type integer?
     opts.buffer = nil ---@type integer?
     for _, m in ipairs(mode) do
       vim.api.nvim_buf_set_keymap(bufnr, m, lhs, rhs, opts)
     end
   else
+    opts.buf = nil
     opts.buffer = nil
```

Comment:
just set `opts.buffer=nil` immediately after line 74. that makes the rest of the code clearer, it will only deal with `opts.buf`.

same comment if there are similar cases in the rest of the PR